### PR TITLE
Added Endpointslice support

### PR DESCRIPTION
**Pull request by jayaramsatya**
1. Added watchers to listen for Endpoint slices changes both in controller and hostagent code
2. Service points to multiple endpointslice Object and each Slice Object points to list of Endpoints
   made changes for the Services code accordingly.
3. Egress netwokpolicy changes are done to handle the Endpoint Slices
4. Currently endpointslice support flag is turned off for this commit
5. Added multiple  UT's in services and networkpolicies by enabling the endpointslice support flag in tests

### CODE REVIEW

File: pkg/controller/network_policy.go [hunk 1]
```diff
@@ -795,71 +811,138 @@ func (cont *AciController) getServiceAugmentBySubnet(subj apicapi.ApicObject,
...
+							proto, port, service)
+
+						logger.WithFields(logrus.Fields{
+							"proto":   proto,
+							"port":    port,
+							"service": servicekey,
+						}).Debug("Allowing egress for service by subnet match")
+					}
+				})
+		}
```
Comment (kishorejonnala) [inline pkg/controller/network_policy.go:943]: Can we move this code into a function? Possibly collapse both cases (endpoint slice and endpoint) into one routine.

Comment (kishorejonnala) [reply pkg/controller/network_policy.go:943]: @jojimt  Pls take a look fron Cloud perspective (ex: populating Region/Zone information)

Comment (jayaramsatya) [reply pkg/controller/network_policy.go:943]: Sure Kishore will take a look

File: pkg/controller/network_policy.go [hunk 2]
```diff
@@ -802,62 +818,116 @@ func (cont *AciController) getServiceAugmentBySubnet(subj apicapi.ApicObject,
...
 		service := serviceobj.(*v1.Service)
-		endpoints := endpointsobj.(*v1.Endpoints)
-
-		for _, svcPort := range service.Spec.Ports {
-			if prs.port != nil &&
-				(svcPort.Protocol != *prs.port.Protocol ||
-					svcPort.TargetPort.String() !=
-						prs.port.Port.String()) {
-				// egress rule does not match service target port
+		if !cont.endPointSliceEnabled {
```
Comment (jojimt) [inline pkg/controller/network_policy.go:825]: Can we avoid this check everywhere in the code? We should define a go interface or function pointer that is setup at init and does the appropriate thing.

File: pkg/controller/services.go [hunk 3]
```diff
@@ -978,18 +1056,50 @@ func (cont *AciController) opflexDeviceDeleted(dn string) {
...
-			key + ": " + err.Error())
-		return
-	}
-
 	aobj := apicapi.NewVmmInjectedSvc(cont.vmmDomainProvider(),
 		cont.config.AciVmmDomain, cont.config.AciVmmController,
 		service.Namespace, service.Name)
 	aobjDn := aobj.GetDn()
 	aobj.SetAttr("guid", string(service.UID))
+	if !cont.endPointSliceEnabled {
```
Comment (jojimt) [inline pkg/controller/services.go:1064]: Let's avoid this if check and use some abstraction so this code doesn't need to know the details of the k8s object.

File: pkg/hostagent/agent.go [hunk 4]
```diff
@@ -138,6 +140,7 @@ func NewHostAgent(config *HostAgentConfig, env Environment, log *logrus.Logger)
 				DefaultNs,
 			},
 		},
+		endPointSliceEnabled: false,
```
Comment (jojimt) [inline pkg/hostagent/agent.go:143]: This should come from config, right? You can default the config to false.

Comment (jayaramsatya) [reply pkg/hostagent/agent.go:143]: Yes Joji, i will change this code reading it from config

File: pkg/controller/network_policy.go (hunk 2 above)
Comment (jayaramsatya) [reply pkg/controller/network_policy.go:825]: Sure will create interface class for all this functions

**Comment by jayaramsatya** [comment]
LGTM  ACC provision changes are still pending once that is merged. we can merge this code.

**Comment by jayaramsatya** [comment]
Addressed some abstraction at the acc controller code.  still some pending code in hostagent will take care of it in next commit.

**Pull request merged**